src/modules/mon/mon.c

```diff
@@ -0,0 +1,416 @@
+            flux_log (ctx->h, LOG_ERR, "%s: flux_rpc: %s", __FUNCTION__,
+                      strerror (errno));
+        }
+        flux_rpc_destroy (rpc);
+        Jput (o);
+    }
+}
+
+static void mon_sink (flux_reduce_t *r, int batchnum, void *arg)
+{
+    ctx_t *ctx = arg;
+    JSON o;
+
+    while ((o = flux_reduce_pop (r))) {
+        JSON oo, data, odata;
+        const char *name;
+        int epoch;
+        char *key = NULL;
+
+        if (!Jget_str (o, "name", &name) || !Jget_int (o, "epoch", &epoch)
+                                         || !Jget_obj (o, "data", &data)) {
+            flux_log (ctx->h, LOG_ERR, "%s: flux_rpc: malformed object",
+                      __FUNCTION__);
+        } else {
+            key = xasprintf ("mon.%s.%d", name, epoch);
+            if (kvs_get (ctx->h, key, &oo) == 0) {
+                if (Jget_obj (oo, "data", &odata))
+                    Jmerge (data, Jget (odata));
+                Jput (oo);
+            }
+            if (kvs_put (ctx->h, key, o) < 0) {
+                flux_log (ctx->h, LOG_ERR, "%s: kvs_put: %s",
+                          __FUNCTION__, strerror (errno));
+            } else if (kvs_commit (ctx->h) < 0) {
+                flux_log (ctx->h, LOG_ERR, "%s: kvs_commit: %s",
+                          __FUNCTION__, strerror (errno));
+            }
+            if (ctx->debug)
+                flux_log (ctx->h, LOG_DEBUG, "sink %s", name);
+        }
+        Jput (o);
+        if (key)
+            free (key);
+    }
+}
+
+/* helper for mon_reduce */
+static JSON pop_item (flux_reduce_t *r, flux_t h, JSON *data)
+{
+    int epoch;
+    const char *name;
+    JSON o;
+
+    if ((o = flux_reduce_pop (r))) {
+        if (!Jget_int (o, "epoch", &epoch)
+                                       || !Jget_str (o, "name", &name)
+                                       || !Jget_obj (o, "data", data)) {
+            flux_log (h, LOG_ERR, "%s: malformed item dropped",
+                      __FUNCTION__);
+            Jput (o);
+            return NULL;
+        }
+    }
+    return o;
+}
+
+static void mon_reduce (flux_reduce_t *r, int batchnum, void *arg)
+{
+    ctx_t *ctx = arg;
+    JSON a, b, a_data, b_data;
+    int count = 0;
+
+    if ((a = pop_item (r, ctx->h, &a_data))) {
+        count++;
+        while ((b = pop_item (r, ctx->h, &b_data))) {
+            Jmerge (a_data, Jget (b_data));
+            Jput (b);
+            count++;
+        }
+        if (flux_reduce_push (r, a) < 0)
+            oom ();
+    }
+    if (ctx->debug)
+        flux_log (ctx->h, LOG_DEBUG, "reduce %d", count);
+}
+
+static struct flux_msghandler htab[] = {
+    { FLUX_MSGTYPE_EVENT,   "hb",             heartbeat_cb },
+    { FLUX_MSGTYPE_REQUEST, "mon.push",       push_request_cb },
+    FLUX_MSGHANDLER_TABLE_END,
+};
+
+int mod_main (flux_t h, int argc, char **argv)
+{
+    ctx_t *ctx = getctx (h);
+
+    if (argc == 1 && !strcmp (argv[0], "debug"))
```

Comment:
We don't need to subscribe to hb event with flux_msg_watcher_addvec any longer? (Sorry, I haven't followed the interface churn for this one lately...)